packages/astro/src/core/render/core.ts

```diff
@@ -97,7 +97,14 @@ export async function render(opts: RenderOptions): Promise<string> {
 		scripts,
 	});
 
-	let html = await renderPage(result, Component, pageProps, null);
+	let html = await renderToString(result, Component, pageProps, null);
+
+	// handle final head injection if it hasn't happened already
+	if (html.indexOf("<!--astro:head:injected-->") == -1) {
+		html = await renderHead(result) + html;
+	}
```

Comment:
Slight deviation from [RFC0007](https://github.com/withastro/rfcs/blob/main/proposals/0007-finalize-head-body.md)... if we haven't injected any head elements yet, the user may not have rendered a `<head>` element so we just prepend them to the file.